### ISSUE TITLE
Training stages 8 requests per layer where 2 would do; ~17 GB read per step

### ISSUE THREAD

**Issue by pjordanandrsn** [pjordanandrsn/grouped-nf4-gemm#58]
## What

A training step stages **8 requests per layer**, where forward + gradient-checkpoint recompute should be 2. Measured on Qwen3-30B-A3B (48 layers, `hot_rows=128`), reading `ColdTier.stats()` deltas per step:

```
step  requests  rows_req     hits   misses  hit_rate   MB read
   0       384     26232    19755     6477     0.753   17191.3
   1       380     22180    16767     5413     0.756   14367.2
   2       380     27612    20885     6727     0.756   17854.9
```

384 / 48 layers = **8 stage requests per layer per step**. Routing touches a median 67 unique experts per layer per forward, and 8 x 48 x 67 ~ 25,700 matches `rows_requested`.

**Each step reads ~17 GB from disk — more than the entire 16.31 GB arena.**

## Why it matters

Disk is the term that does *not* shrink when the GPU gets faster. Measured across two architectures: going RTX 3090 -> L40S, the host arm's step time nearly halved (1.938 -> 0.977 s on OLMoE) while the arena arm improved only 1.19-1.41x, because its NVMe reads are indifferent to the card. So the arena's *relative* cost grows on faster hardware — 1.331 -> 2.248 on OLMoE, 1.238 -> 1.708 on Qwen3.

Cutting the request amplification attacks exactly that term.

## What is NOT the problem

- **`hot_rows` sizing.** Sweeping 128 / 384 / 1024 produced no resolvable step-time difference (ranges overlap; control `host_self` 0.982 [0.948-1.144]). Raising it costs load time (0.148 -> 0.216 vs host) and ~8x the pinned RAM for nothing measurable.
- **Queue depth.** `qd=4` is already at link saturation per N0.
- **Eviction policy quality.** Hit rate is 75.5%, but with `hot_rows=128` the tier holds exactly one layer's worth, so advancing a layer evicts everything behind it — 18,489 evictions over 3 steps. That is a capacity consequence, not an LFU failure.

The 17 GB/step is largely overlapped with compute (`O_DIRECT`, qd 4), which is why the `hot_rows` sweep came back null — the bytes are real but partly hidden. That makes this a throughput headroom question rather than a visible stall, and it will surface as hardware gets faster.

## To find out

1. **Where do the 8 come from?** Candidates: one `ensure()` per projection rather than per expert set; the `weights_fn` closure re-reading per backward invocation; `enable_fast_train(dgrad=True)` adding passes. Instrument `ColdTier.ensure` call sites with a traceback tally over one step — this is a counter question, not an experiment.
2. **How much is redundant?** If the same (layer, expert) set is requested N times per step, requests 2..N should be pure hits. They largely are (75.5%), so the amplification costs *request overhead and eviction pressure* more than raw re-reads — worth quantifying before optimising.
3. **Does coalescing help?** If several requests per layer carry the same row set, merging them into one `ensure()` per layer per pass is a local change.

## Provenance

Measured 2026-08-13 on the QNAP A2000 with published wheels e4b 0.17.4 / gnf4 0.9.0, `hot_rows=128`, seq 384, 3 steps, gradient checkpointing on (required by the arena path). Counters only — no wall-clock claim is made from this box.

**Comment by pjordanandrsn** [comment]
Closing, and the docs that still listed this as open are corrected in #381 (`75e95fa`). The trace in this thread answered it: six of the eight calls per layer are cache hits (0.02 s/step), and the ~17 GB read per step is the routing floor at \`hot_rows=128\`. The real training-step stall is #60, still open. Triage pass 2026-09-23.
